packages/plugin-session-replay/src/session-replay.ts

```diff
@@ -0,0 +1,286 @@
+import { AMPLITUDE_PREFIX, BaseTransport } from '@amplitude/analytics-core';
+import { BrowserConfig, EnrichmentPlugin, Event, PluginType, Status } from '@amplitude/analytics-types';
+import * as IDBKeyVal from 'idb-keyval';
+import { pack, record } from 'rrweb';
+import { shouldSplitEventsList } from './helpers';
+import { MAX_RETRIES_EXCEEDED_MESSAGE, STORAGE_FAILURE, SUCCESS_MESSAGE, UNEXPECTED_ERROR_MESSAGE } from './messages';
+import { Events, IDBStore, SessionReplayContext } from './typings/session-replay';
+
+const SESSION_REPLAY_SERVER_URL = 'https://api-secure.amplitude.com/sessions/track';
+const STORAGE_PREFIX = `${AMPLITUDE_PREFIX}_replay_unsent`;
+const PAYLOAD_ESTIMATED_SIZE_IN_BYTES_WITHOUT_EVENTS = 200; // derived by JSON stringifying an example payload without events
+const MAX_EVENT_LIST_SIZE_IN_BYTES = 20 * 1000000 - PAYLOAD_ESTIMATED_SIZE_IN_BYTES_WITHOUT_EVENTS;
+
+export class SessionReplayPlugin implements EnrichmentPlugin {
+  name = '@amplitude/plugin-session-replay';
+  type = PluginType.ENRICHMENT as const;
+  // this.config is defined in setup() which will always be called first
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore
+  config: BrowserConfig;
+  storageKey = '';
+  retryTimeout = 1000;
+  events: Events = [];
+  currentSequenceId = 0;
+  private scheduled: ReturnType<typeof setTimeout> | null = null;
+  queue: SessionReplayContext[] = [];
+  stopRecordingEvents: ReturnType<typeof record> | null = null;
+  maxPersistedEventsSize = MAX_EVENT_LIST_SIZE_IN_BYTES;
+
+  async setup(config: BrowserConfig) {
+    config.loggerProvider.log('Installing @amplitude/plugin-session-replay.');
+
+    this.config = config;
+    this.storageKey = `${STORAGE_PREFIX}_${this.config.apiKey.substring(0, 10)}`;
+    void this.emptyStoreAndReset();
+  }
+
+  execute(event: Event) {
+    event.event_properties = {
+      ...event.event_properties,
+      session_replay_enabled: true,
```

Comment:
Are we sure we want to modify the `event properties` directly? This can cause collisions and override customer defined values. I think `session_replay_enabled` is fairly unique but still want to call this out.

If we are ok modifying the event properties directly, it might be good to create some sort of namespacing e.g. `amp_session_replay_enabled` so we can warn customers not to create properties that start with `amp_` in order to prevent data loss.

Reply:
Yep after a team discussion we decided to rename this property to `[Amplitude] Session Recorded` to keep it inline with the other default properties we already have